service/worker/batcher/workflow.go

```diff
@@ -20,7 +20,328 @@
+func BatchWorkflow(ctx workflow.Context, batchParams BatchParams) error {
+	batchParams, err := setDefaultParams(batchParams)
+	if err != nil {
+		return err
+	}
+	batchActivityOptions.HeartbeatTimeout = batchParams.ActivityHeartBeatTimeout
+	opt := workflow.WithActivityOptions(ctx, batchActivityOptions)
+	return workflow.ExecuteActivity(opt, batchActivityName, batchParams).Get(ctx, nil)
+}
+
+func setDefaultParams(params BatchParams) (BatchParams, error) {
+	if params.BatchType == "" || params.Reason == "" || params.DomainName == "" || params.QueryCondition == "" {
+		return BatchParams{}, fmt.Errorf("must provide required parameters: BatchType/Reason/DomainName/QueryCondition")
+	}
+	if params.RPS <= 0 {
+		params.RPS = defaultRPS
+	}
+	if params.Concurrency <= 0 {
+		params.Concurrency = defaultConcurrency
+	}
+	if params.AttemptsOnRetryableError <= 0 {
+		params.AttemptsOnRetryableError = defaultAttemptsOnRetryableError
+	}
+	if params.ActivityHeartBeatTimeout <= 0 {
+		params.ActivityHeartBeatTimeout = defaultActivityHeartBeatTimeout
+	}
+	switch params.BatchType {
+	case BatchTypeReset:
+		fallthrough
+	default:
+		return BatchParams{}, fmt.Errorf("not supported batch type: %v", params.BatchType)
+	case BatchTypeTerminate:
+
+	}
+	return params, nil
+}
+
+func BatchActivity(ctx context.Context, batchParams BatchParams) error {
+	batcher := ctx.Value(batcherContextKey).(*Batcher)
+
+	hbd := HeartBeatDetails{}
+	startOver := true
+	if activity.HasHeartbeatDetails(ctx) {
+		if err := activity.GetHeartbeatDetails(ctx, &hbd); err == nil {
+			startOver = false
+		} else {
+			//TODO error metrics
+			getActivityLogger(ctx).Error("Failed to recover from last heartbeat, start over from beginning", tag.Error(err))
+		}
+	}
+
+	if startOver {
+		resp, err := batcher.svcClient.CountWorkflowExecutions(ctx, &shared.CountWorkflowExecutionsRequest{
+			Domain: common.StringPtr(batchParams.DomainName),
+			Query:  common.StringPtr(batchParams.QueryCondition),
+		})
+		if err != nil {
+			return err
+		}
+		hbd.TotalEstimate = resp.GetCount()
+	}
+	rateLimiter := rate.NewLimiter(rate.Limit(batchParams.RPS), batchParams.RPS)
+	taskCh := make(chan taskDetail, pageSize)
+	var succCount, errCount int32
+	for i := 0; i < batchParams.Concurrency; i++ {
+		go startTaskProcessor(ctx, batchParams, taskCh, rateLimiter, &succCount, &errCount)
+	}
+
+	for {
+		// TODO https://github.com/uber/cadence/issues/2154
+		//  Need to improve scan concurrency because it will hold a ES resource for a long time.
+		//  we can't use list API because terminate / reset will mutate the result.
+		resp, err := batcher.svcClient.ScanWorkflowExecutions(ctx, &shared.ListWorkflowExecutionsRequest{
+			PageSize:      common.Int32Ptr(int32(pageSize)),
+			Domain:        common.StringPtr(batchParams.DomainName),
+			NextPageToken: hbd.PageToken,
+			Query:         common.StringPtr(batchParams.QueryCondition),
+		})
+		if err != nil {
```

Comment:
Since it is using service client, I think we need retry to handle transient internal server error (such as service busy error) here instead of activity retry.

Reply:
OK